src/Microsoft.ML.Data/Transforms/ValueMappingTransform.cs

```diff
@@ -0,0 +1,515 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Data;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.CommandLine;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Data.IO;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Transforms;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+[assembly: LoadableClass(ValueMappingTransform.Summary, typeof(IDataTransform), typeof(ValueMappingTransform), null, typeof(SignatureLoadDataTransform),
+    "Value Mapping Transform", ValueMappingTransform.LoaderSignature)]
+
+[assembly: LoadableClass(ValueMappingTransform.Summary, typeof(ValueMappingTransform), null, typeof(SignatureLoadModel),
+    "Value Mapping Transform", ValueMappingTransform.LoaderSignature)]
+
+namespace Microsoft.ML.Transforms
+{
+    public sealed class ValueMappingEstimator<TKeyType, TValueType> : TrivialEstimator<ValueMappingTransform<TKeyType, TValueType>>
+    {
+        private (string input, string output)[] _columns;
+
+        public ValueMappingEstimator(IHostEnvironment env, IEnumerable<TKeyType> keys, IEnumerable<TValueType> values, params (string input, string output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingEstimator<TKeyType, TValueType>)),
+                    new ValueMappingTransform<TKeyType, TValueType>(env, keys, values, columns))
+        {
+            _columns = columns;
+        }
+
+        public ValueMappingEstimator(IHostEnvironment env, IEnumerable<TKeyType[]> keys, IEnumerable<TValueType> values, params (string input, string output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingEstimator<TKeyType, TValueType>)),
+                    new ValueMappingTransform<TKeyType, TValueType>(env, keys, values, columns))
+        {
+            _columns = columns;
+        }
+
+        public ValueMappingEstimator(IHostEnvironment env, IEnumerable<TKeyType> keys, IEnumerable<TValueType[]> values, params (string input, string output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingEstimator<TKeyType, TValueType>)),
+                    new ValueMappingTransform<TKeyType, TValueType>(env, keys, values, columns))
+        {
+            _columns = columns;
+        }
+
+        public ValueMappingEstimator(IHostEnvironment env, IEnumerable<TKeyType[]> keys, IEnumerable<TValueType[]> values, params (string input, string output)[] columns)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(nameof(ValueMappingEstimator<TKeyType, TValueType>)),
+                    new ValueMappingTransform<TKeyType, TValueType>(env, keys, values, columns))
+        {
+            _columns = columns;
+        }
+
```

Comment:
they seem all identical to me?

Reply:
They are different based upon an [] argument which allows for a vector type to be used as either a key or a value. So there is one constructor for each scenario:
1) key->value
2) key[]->value
3) key->value[]
4) key[]->value[]

I would have rather have TKeyType->TValueType, and have the TKeyType or TValueType be an array type or single type which would reduce this to one constructor -- but this caused other issues for code that deals with types or has generic functions. 

---
In reply to: [236478883](https://github.com/dotnet/machinelearning/pull/1710#discussion_r236478883) [](ancestors = 236478883)